### ISSUE TITLE
Is RIA backup storage necessary for a HPC dataset?

### ISSUE THREAD

**Issue by anikfal** [datalad/datalad#7156]
My main issue is to make the dataset available to collaborators. I don't need any backup, and actually it's not possible for terabytes of data.

It seems RIA makes a backup of the original dataset (how can I avoid this?).

Also by applying `--storage-sibling=off`, the dataset cannot be retrieved by `datalad get`, neither after cloning from the server (`datalad clone ria+ssh ...`), nor after cloning from Gitlab.

**Comment by yarikoptic** [comment]
can you make your original dataset available as a git repository over ssh or over http ?  it could actually be some "dedicated" ephemeral (symlinked `.git/annex`) clone which you expose to them while sharing `.git/annex/objects`.

**Comment by anikfal** [comment]
> can you make your original dataset available

https://gitlab.jsc.fz-juelich.de/nikfal1/test_detect.git

I didn't create any RIA for the subdataset inside it.

**Comment by yarikoptic** [comment]
I would leave it to other @datalad/developers from Juelich to follow up since they would know the situation better ;-)

**Comment by bpoldrack** [comment]
@anikfal 

1. No RIA is necessary to begin with. Whether it is useful, depends on how you want to use your datasets.
2. Generally, if you use a RIA store, it's a misconception to think of it as a backup. The idea would be to make the RIA store the "original" place, where compute jobs or collaborators would clone from and possibly push back to. The datasets in the store would be the thing to share. Also note: What's "making a backup" here is not `create-sibling-ria`, but you `datalad push`'ing things there.
3. Independently on that, you can use the RIA concept for a backup, too. But there's no such requirement. A backup could rely on the storage provider's (JSC in your case, I guess) backup routine for the location you have the store at.
4.  > It seems RIA makes a backup of the original dataset (how can I avoid this?).
     
     You kind of answered this yourself. If you create a RIA it would by default take the git history and the data content if you `datalad push` to it. By setting ` --storage-sibling=off` you set the dataset in the store up to take no data. Apart from the basic setup of the RIA sibling, you can determine what goes into the store by specifying what you `push` to it. See `datalad push --help`.
     But again: I think the main misconception here is, that you look at your original dataset that you created the RIA sibling from as the central piece that would need to be shared, rather than the created RIA sibling to be that. Note, that if your "original" dataset is the piece that actually holds the data, sharing would require collaborators to have access to that. It's likely easier to manage access for such a store as it's a dedicated location for this purpose. What ties into that aspect is, that compute jobs will need access as well. So, the data source (whether you decide for RIA or your "original") needs to be with the JSC and accessible from the compute nodes.
     Is there a reason you'd rather share the data from the original location than the RIA store?
5.  > Also by applying --storage-sibling=off, the dataset cannot be retrieved by datalad get
    I'm not sure what you are asking here. Yes, if the store has no storage-sibling it can not have data content. Hence, a clone that only knows this sibling can not retrieve any data. If the only other location is GitLab (as it seems to be your setup) then no data source is known and accessible by such a clone.
    

To sum up: The general idea would be to make the RIA store the "original" place to share, that collaborators and compute jobs can clone from (the detour via GitLab is likely nice for human collaborators, while you probably want the store to be with the JSC and have compute jobs clone directly from it) and get (only) the data they need. When done and possible changes to the dataset are pushed back into the store, those clones can either be wiped completely or simply drop the data again to not hold a local copy of it anymore.

**Comment by bpoldrack** [comment]
There may be a good reason to not put the data into the store and leave it where it already is. (Would be good to know what is that reason.) In that case, you need to make this "original" location known and accessible to collaborators. Essentially the same way you made the RIA location known to users cloning from GitLab. How exactly that would need to be done depends on what that location is and how people would have access to it. One of the (minor) reasons to use a RIA store is, that datalad can dynamically access it via local file system operation when on the same system as well as via SSH (and even HTTP(S) if you can serve the store with a webserver).

**Comment by anikfal** [comment]
> The general idea would be to make the RIA store the "original" place to share

Thanks for the elaboration. However, I'm currently not sure how to implement this idea.

**Comment by anikfal** [comment]
@bpoldrack 

In the original dataset, after `datalad puth --to ria_sibling` and creating the backup, I can `datalad drop ...` the content of the original dataset. So the original dataset will be empty (filled with only some broken symlinks), while the backup data is available in RIA, and can be retrieved by `datalad clone ria+ssh ...`. No duplication in this case.

I think this might be a workaround I can do for this issue.

**Comment by bpoldrack** [comment]
Yes, this is exactly the idea. "Move" the data instead of just pushing to copy location, if you do not want a redundant copy. And "move" means: push and drop.

**Comment by anikfal** [comment]
This is what I want in case of sharing data from the server to local systems. However, there will be a critical issue:
Now that all data are stored in RIA, and the contents of the original dataset have been dropped, what if some other users in the same HPC server want to access data? Each user has to retrieve the data as a copy (`datalad clone ria+file ...`). They cannot access to the original data as links, since the contents of the original dataset have been dropped, leaving some broken symlinks.
Is there any workaround for this issue?

I wish it could be possible to directly publish the original dataset in a HPC system to the local systems or Gitlab, without making any RIA at all.

**Comment by bpoldrack** [comment]
> I wish it could be possible to directly publish the original dataset in a HPC system to the local systems or Gitlab, without making any RIA at all.

In principle that absolutely is possible, but how exactly very much depends on what the infrastructure allows you to do. 

If you need data access for collaborators on the HPC only, then you don't need any of that - they can just clone from the path your original dataset is at. Redundant copies can be avoided, but I'd strongly suggest to discuss this in office hour, as it does not fit every usecase. (You want to make extra sure, that the content cannot be changed from within clones - otherwise you sabotage your version control) Keyword to look up: annex thin mode.
If collaborators outside need data access, however, you'd need to serve the original dataset somehow. RIA helps with that in so far, as it provides SSH access. If you can set up a git and git-annex aware webserver, this would also be an option. But I don't think you can usually do that on HPC infrastructure - it's typically not meant as a data storage provider.
There is another aspect to consider: You may (eventually) need to have a bunch of subdatasets. In a RIA store those can be easily discovered. Without it you would need to make sure that the superdataset also has accessible URLs for the subdatasets recorded. RIA does that for you in some sense.

> Each user has to retrieve the data as a copy [...] They cannot access to the original data as links [...] Is there any workaround for this issue?

If the data is in a RIA store, users/compute jobs could clone via `datalad clone --reckless ephemeral "ria+file: ....`. However, the option is called `reckless` for a reason - if used wrong it could mess with the original data. This should only be used with "proven", scripted workflows.

Generally, I do wonder whether your desire to not have copies is well advised. Do collaborators need a full copy of everything? The idea here would be to only ever `datalad get` the files needed for a particular analysis, `datalad save` the results (likely in a derived dataset), push the results to store and then drop everything again. So in most cases that would "only" be temporary, partial copies.
Obviously that may still be too much, but I don't know - can reasonably estimate what people would actually need at a time?

**Comment by anikfal** [comment]
@bpoldrack 
Thanks for the elaboration. I need to be more familiar with git-annex.

> If the data is in a RIA store, users/compute jobs could clone via `datalad clone --reckless ephemeral "ria+file: ....`.

This clone-command quickly solves the issue.
